File: Easy/PHP/GenerateDocument.php
<?php 

namespace Easy\PHP\GenerateDocument;

function generate_document2($characters, $document)
{
    $charsCount = [];
    $chars = str_split($characters);
    $docs = str_split($document);

    foreach ($chars as $character) {
        if (in_array($character, $charsCount)) {
            $charsCount[$character] = 0;
        }
        $charsCount[$character]++;
    }
    

    foreach ($docs as $character) {
        if (in_array($character, $charsCount) || $charsCount[$character] == 0) {
            return false;
        }
        $charsCount[$character]--;
    }
    return true;
}


function generate_document1($characters, $document) {
    for ($i=0; $i< strlen($document); $i++) {
        $documentCount = charsFrequency($document[$i], $document);
        $charactersCount = charsFrequency($document[$i], $characters);
    }

    if ($documentCount > $charactersCount) {
        return false;
    }

    return true;
}

function charsFrequency($char, $toScan)
{
    $count = 0;
    for ($i=0; $i<strlen($toScan); $i++) {
        if ($toScan[$i] == $char) {
            $count ++;
        }
    }

    return $count;
}


$characters = "heeelloworlddd!";
$document = "dll2row!oleh";
$result = generate_document2($characters, $document);

var_dump($result);

File: Easy/PYTHON/GenerateDocument.py
def generate_document2(characters, document):
    charsCount = {}

    for character in characters:
        if character not in charsCount:
            charsCount[character] = 0

        charsCount[character] += 1

    for character in document:
        if character not in charsCount or charsCount[character] == 0:
            return False

        charsCount[character] -= 1

    return True


def generate_document1(characters, document):
    for i in range(len(document)):
        documentCount = chars_frequency(document[i], document)
        charactersCount = chars_frequency(document[i], characters)

    if(documentCount > charactersCount):
        return False

    return True


def chars_frequency(char, toScan):
    count = 0
    for i in range(len(toScan)):
        if toScan[i] == char:
            count += 1

    return count


characters = "heeelloworlddd!"
document = "dllrow!oleh"
result = generate_document2(characters, document)
print(result)
